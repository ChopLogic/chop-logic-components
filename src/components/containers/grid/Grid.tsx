import { useState } from 'react';
import { useElementIds } from 'hooks/use-element-ids';

import ChopLogicGridBody from './elements/Body';
import GridColumnGroup from './elements/ColumnGroup';
import ChopLogicGridHead from './elements/Head';
import { StyledGrid, StyledGridCaption } from './Grid.styled';
import { ChopLogicGridProps } from './types';

const ChopLogicGrid: React.FC<ChopLogicGridProps> = ({ columns, data, id, onSelect, renderDataItem, caption, selectable = false }) => {
  const { elementId } = useElementIds(id);
  const [selectedIds, setSelectedIds] = useState<string[]>([]);
  const allIds = data.map((item) => item.id);
  const isAllSelected = allIds.length === selectedIds.length;
  const isAllCheckboxDisabled = data.some((item) => item?.disabled);

  const handleSelect = (ids: string[]) => {
    setSelectedIds(ids);
    onSelect?.(ids);
  };

  const handleSelectAll = () => {
    handleSelect(allIds);
  };

  const handleDeselectAll = () => {
    handleSelect([]);
  };

  const handleSelectRowById = (id: string) => {
    const newIds = [...selectedIds, id];
    handleSelect(newIds);
  };

  const handleDeselectRowById = (id: string) => {
    const newIds = selectedIds.filter((item) => item !== id);
    handleSelect(newIds);
  };

  return (
    <StyledGrid>
      {caption && <StyledGridCaption>{caption}</StyledGridCaption>}
      <GridColumnGroup columns={columns} selectable={selectable} />
      <ChopLogicGridHead
        gridId={elementId}
        columns={columns}
        selectable={selectable}
        selectAll={handleSelectAll}
        deselectAll={handleDeselectAll}
        isAllSelected={isAllSelected}
        isAllCheckboxDisabled={isAllCheckboxDisabled}
      />
      <ChopLogicGridBody
        columns={columns}
        data={data}
        selectable={selectable}
        selectRowById={handleSelectRowById}
        deselectRowById={handleDeselectRowById}
        selectedIds={selectedIds}
        renderDataItem={renderDataItem}
      />
    </StyledGrid>
  );
};

export default ChopLogicGrid;
