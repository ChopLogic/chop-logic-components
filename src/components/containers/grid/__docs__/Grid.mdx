import { Meta, Story, Canvas } from '@storybook/blocks';
import * as ChopLogicGridStories from './Grid.stories';

<Meta of={ChopLogicGridStories} title='Grid' />

# ChopLogicGrid

The `ChopLogicGrid` component is a styled grid component built on top of the default HTML `table` element.

#### Example

<Canvas>
  <Story of={ChopLogicGridStories.DefaultGrid} />
</Canvas>

## Usage

```tsx
import React from 'react';
import ChopLogicGrid from 'chop-logic-components';
import { DATA, COLUMS } from './constants';

const ExampleComponent = () => <ChopLogicGrid selectable={false} data={DATA} columns={COLUMNS} />;

export default ExampleComponent;
```

### Props

_Required props_

- **data** `ChopLogicGridItem[]` - An array of items to be displayed in the grid. Each item is a JavaScript object with a mandatory `id` field.
- **columns** `ChopLogicGridColumn[]` - An array of objects that describes the grid columns. Each column must contain `field` property that points to the value in the data item, and `title` property that describes the column title.

_Optional props_

- **selectable** `boolean` - A value that indicates whether the grid has a column with checkboxes.
- **caption** `string` - An optional title of the grid.
- **onSelect** `(ids: string[]) => void` - An optional callback that is called when one or more grid rows are selected or deselected.
- **renderDataItem** `RenderDataItemCallback` - An optional callback that takes a data item and a column field as parameters and returns a JSX element. This callback is required for custom rendering of grid cells.
- **HTMLAttributes** - some global HTML attributes: **className**, **style**, **tabIndex**, etc.

### Types

```ts
type ChopLogicGridProps = {
  columns: ChopLogicGridColumn[];
  data: ChopLogicGridItem[];
  caption?: string;
  id?: string;
  selectable?: boolean;
  tabIndex?: number;
  className?: string;
  style?: React.CSSProperties;
  renderDataItem?: RenderDataItemCallback;
  onSelect?: (ids: string[]) => void;
};
```

```ts
type ChopLogicGridColumn = {
  field: string;
  title?: string;
  component?: React.ReactElement;
  highlighted?: boolean;
  className?: string;
};
```

```ts
type ChopLogicGridItem = {
  id: string;
  disabled?: boolean;
  [key: string]: unknown;
};
```

```ts
export type RenderDataItemCallback = (item: ChopLogicGridItem, field: string) => JSX.Element;
```
