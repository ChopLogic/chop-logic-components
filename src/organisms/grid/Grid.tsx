import { GridProps } from '@models';
import { getClassName } from '@utils';
import { FC } from 'react';

import { GridBody } from './body/GridBody';
import { GridColumnGroup } from './column-group/GridColumnGroup';
import { useGridController } from './Grid.controller';
import styles from './Grid.module.scss';
import { GridHead } from './head/GridHead';

const Grid: FC<GridProps> = ({ columns, data, id, onSelect, renderDataItem, caption, selectable = false, className, ...rest }) => {
  const {
    elementId,
    handleSelectAll,
    handleDeselectAll,
    isAllCheckboxDisabled,
    isAllSelected,
    handleDeselectRowById,
    handleSelectRowById,
    selectedIds,
  } = useGridController({ id, data, onSelect });

  return (
    <table {...rest} className={getClassName([styles.grid, className])}>
      {caption && <caption className={styles.grid_caption}>{caption}</caption>}
      <GridColumnGroup columns={columns} selectable={selectable} />
      <GridHead
        gridId={elementId}
        columns={columns}
        selectable={selectable}
        selectAll={handleSelectAll}
        deselectAll={handleDeselectAll}
        isAllSelected={isAllSelected}
        isAllCheckboxDisabled={isAllCheckboxDisabled}
      />
      <GridBody
        columns={columns}
        data={data}
        selectable={selectable}
        selectRowById={handleSelectRowById}
        deselectRowById={handleDeselectRowById}
        selectedIds={selectedIds}
        renderDataItem={renderDataItem}
      />
    </table>
  );
};

export default Grid;
