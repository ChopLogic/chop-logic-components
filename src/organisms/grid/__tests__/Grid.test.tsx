import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { describe, expect, it, vi } from 'vitest';

import Grid from '../Grid';

describe('Grid', () => {
  const testProps = {
    id: 'grid-id',
    columns: [
      { title: 'Company', field: 'company' },
      { title: 'Contact', field: 'contact' },
      { title: 'Country', field: 'country' },
      { title: 'Phone', field: 'phone' },
    ],
    data: [
      { id: 'row1', company: '', contact: 'Maria', country: 'Germany', phone: '123-345-7890' },
      { id: 'row2', company: '', contact: 'Francisco', country: 'Mexico', phone: '123-345-7890' },
      { id: 'row3', company: 'Ernst Handel', contact: 'Roland Mendel', country: 'Austria', phone: '123-345-7890' },
      { id: 'row4', company: 'Island Trading', contact: 'Helen Bennett', country: 'UK', phone: '123-345-7890' },
      { id: 'row5', company: 'Laughing Bacchus', contact: '', country: 'Canada', phone: '123-345-7890' },
      { id: 'row6', company: '', contact: 'Giovanni', country: 'Italy', phone: '123-345-7890' },
    ],
    className: 'test-class',
    style: { width: 100 },
    tabIndex: -1,
  };

  it('should match the snapshot', () => {
    const { asFragment } = render(<Grid {...testProps} selectable />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should not render the checkbox column if the grid is not selectable', () => {
    render(<Grid {...testProps} selectable={false} />);
    expect(screen.queryAllByRole('checkbox')).toHaveLength(0);
  });

  it('should render an optional caption', () => {
    render(<Grid {...testProps} selectable={false} caption='Test Caption' />);
    expect(screen.getByText('Test Caption')).toBeInTheDocument();
  });

  it('should call onSelect() handler when a checkbox is clicked', async () => {
    const mockSelect = vi.fn();
    render(<Grid {...testProps} selectable caption='Test Caption' onSelect={mockSelect} />);
    const firstCheckbox = screen.getAllByRole('checkbox')[0];
    const secondCheckbox = screen.getAllByRole('checkbox')[1];
    await userEvent.click(firstCheckbox);
    expect(mockSelect).toHaveBeenCalledWith(['row1', 'row2', 'row3', 'row4', 'row5', 'row6']);
    await userEvent.click(firstCheckbox);
    expect(mockSelect).toHaveBeenCalledWith([]);
    await userEvent.click(secondCheckbox);
    expect(mockSelect).toHaveBeenCalledWith(['row1']);
  });
});
