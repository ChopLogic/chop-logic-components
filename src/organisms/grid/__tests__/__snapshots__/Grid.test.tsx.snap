// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Grid > should match the snapshot 1`] = `
<DocumentFragment>
  <table
    class="_grid_384728 test-class"
    style="width: 100px;"
    tabindex="-1"
  >
    <colgroup>
      <col
        class="_column_08065d _column__first_08065d"
      />
      <col
        class="_column_08065d"
      />
      <col
        class="_column_08065d"
      />
      <col
        class="_column_08065d"
      />
      <col
        class="_column_08065d"
      />
    </colgroup>
    <thead
      class="_head_7f4cec"
    >
      <tr
        class="_head_row_7f4cec"
      >
        <th
          class="_headerCell_d2069a"
        >
          <div
            class="_checkbox_a434a4"
          >
            <input
              class="_checkbox_input_a434a4"
              id="select_all_grid-id"
              type="checkbox"
            />
            <label
              class="_label_c20562"
              for="select_all_grid-id"
            >
              <svg
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 16 16"
                width="1em"
              >
                <path
                  d="M14 0H2C.9 0 0 .9 0 2v12c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V2c0-1.1-.9-2-2-2zm0 14H2V2h12v12z"
                  fill="currentColor"
                />
              </svg>
            </label>
          </div>
        </th>
        <th
          class="_headerCell_37da90"
        >
          Company
        </th>
        <th
          class="_headerCell_37da90"
        >
          Contact
        </th>
        <th
          class="_headerCell_37da90"
        >
          Country
        </th>
        <th
          class="_headerCell_37da90"
        >
          Phone
        </th>
      </tr>
    </thead>
    <tbody
      class="_body_0a5981"
    >
      <tr
        class="_row_2c96fd"
      >
        <td
          class="_cell_4fae7d"
        >
          <div
            class="_checkbox_a434a4"
          >
            <input
              class="_checkbox_input_a434a4"
              id="select_row_row1"
              type="checkbox"
            />
            <label
              class="_label_c20562"
              for="select_row_row1"
            >
              <svg
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 16 16"
                width="1em"
              >
                <path
                  d="M14 0H2C.9 0 0 .9 0 2v12c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V2c0-1.1-.9-2-2-2zm0 14H2V2h12v12z"
                  fill="currentColor"
                />
              </svg>
            </label>
          </div>
        </td>
        <td
          class="_cell_ae74a7"
        />
        <td
          class="_cell_ae74a7"
        >
          Maria
        </td>
        <td
          class="_cell_ae74a7"
        >
          Germany
        </td>
        <td
          class="_cell_ae74a7"
        >
          123-345-7890
        </td>
      </tr>
      <tr
        class="_row_2c96fd"
      >
        <td
          class="_cell_4fae7d"
        >
          <div
            class="_checkbox_a434a4"
          >
            <input
              class="_checkbox_input_a434a4"
              id="select_row_row2"
              type="checkbox"
            />
            <label
              class="_label_c20562"
              for="select_row_row2"
            >
              <svg
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 16 16"
                width="1em"
              >
                <path
                  d="M14 0H2C.9 0 0 .9 0 2v12c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V2c0-1.1-.9-2-2-2zm0 14H2V2h12v12z"
                  fill="currentColor"
                />
              </svg>
            </label>
          </div>
        </td>
        <td
          class="_cell_ae74a7"
        />
        <td
          class="_cell_ae74a7"
        >
          Francisco
        </td>
        <td
          class="_cell_ae74a7"
        >
          Mexico
        </td>
        <td
          class="_cell_ae74a7"
        >
          123-345-7890
        </td>
      </tr>
      <tr
        class="_row_2c96fd"
      >
        <td
          class="_cell_4fae7d"
        >
          <div
            class="_checkbox_a434a4"
          >
            <input
              class="_checkbox_input_a434a4"
              id="select_row_row3"
              type="checkbox"
            />
            <label
              class="_label_c20562"
              for="select_row_row3"
            >
              <svg
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 16 16"
                width="1em"
              >
                <path
                  d="M14 0H2C.9 0 0 .9 0 2v12c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V2c0-1.1-.9-2-2-2zm0 14H2V2h12v12z"
                  fill="currentColor"
                />
              </svg>
            </label>
          </div>
        </td>
        <td
          class="_cell_ae74a7"
        >
          Ernst Handel
        </td>
        <td
          class="_cell_ae74a7"
        >
          Roland Mendel
        </td>
        <td
          class="_cell_ae74a7"
        >
          Austria
        </td>
        <td
          class="_cell_ae74a7"
        >
          123-345-7890
        </td>
      </tr>
      <tr
        class="_row_2c96fd"
      >
        <td
          class="_cell_4fae7d"
        >
          <div
            class="_checkbox_a434a4"
          >
            <input
              class="_checkbox_input_a434a4"
              id="select_row_row4"
              type="checkbox"
            />
            <label
              class="_label_c20562"
              for="select_row_row4"
            >
              <svg
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 16 16"
                width="1em"
              >
                <path
                  d="M14 0H2C.9 0 0 .9 0 2v12c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V2c0-1.1-.9-2-2-2zm0 14H2V2h12v12z"
                  fill="currentColor"
                />
              </svg>
            </label>
          </div>
        </td>
        <td
          class="_cell_ae74a7"
        >
          Island Trading
        </td>
        <td
          class="_cell_ae74a7"
        >
          Helen Bennett
        </td>
        <td
          class="_cell_ae74a7"
        >
          UK
        </td>
        <td
          class="_cell_ae74a7"
        >
          123-345-7890
        </td>
      </tr>
      <tr
        class="_row_2c96fd"
      >
        <td
          class="_cell_4fae7d"
        >
          <div
            class="_checkbox_a434a4"
          >
            <input
              class="_checkbox_input_a434a4"
              id="select_row_row5"
              type="checkbox"
            />
            <label
              class="_label_c20562"
              for="select_row_row5"
            >
              <svg
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 16 16"
                width="1em"
              >
                <path
                  d="M14 0H2C.9 0 0 .9 0 2v12c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V2c0-1.1-.9-2-2-2zm0 14H2V2h12v12z"
                  fill="currentColor"
                />
              </svg>
            </label>
          </div>
        </td>
        <td
          class="_cell_ae74a7"
        >
          Laughing Bacchus
        </td>
        <td
          class="_cell_ae74a7"
        />
        <td
          class="_cell_ae74a7"
        >
          Canada
        </td>
        <td
          class="_cell_ae74a7"
        >
          123-345-7890
        </td>
      </tr>
      <tr
        class="_row_2c96fd"
      >
        <td
          class="_cell_4fae7d"
        >
          <div
            class="_checkbox_a434a4"
          >
            <input
              class="_checkbox_input_a434a4"
              id="select_row_row6"
              type="checkbox"
            />
            <label
              class="_label_c20562"
              for="select_row_row6"
            >
              <svg
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 16 16"
                width="1em"
              >
                <path
                  d="M14 0H2C.9 0 0 .9 0 2v12c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V2c0-1.1-.9-2-2-2zm0 14H2V2h12v12z"
                  fill="currentColor"
                />
              </svg>
            </label>
          </div>
        </td>
        <td
          class="_cell_ae74a7"
        />
        <td
          class="_cell_ae74a7"
        >
          Giovanni
        </td>
        <td
          class="_cell_ae74a7"
        >
          Italy
        </td>
        <td
          class="_cell_ae74a7"
        >
          123-345-7890
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
